### PR TITLE
[#2525] fix(catalog-hive): Fix classNotFoundError in the Caffeine removal listener

### PR DESCRIPTION
**Pull request by yuqi1129**
### What changes were proposed in this pull request?

Close `HiveClientPool` before closing the class loader as the removal listener is asynchronous and can't make this guarantee.

### Why are the changes needed?

It's bugs that need to be fixed.

Fix: #2525 

### Does this PR introduce _any_ user-facing change?

N/A.

### How was this patch tested?

Test locally.

### CODE REVIEW

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java [hunk 1]
```diff
@@ -136,6 +147,13 @@ private static ThreadFactory newDaemonThreadFactory() {
 
   public void close() {
     clientPoolCache.invalidateAll();
+
+    // Let the remove listener to close the HiveClientPool first then close the class loader.
+    try {
+      Thread.sleep(500);
+    } catch (Exception e) {
+      // Ignore
+    }
```
Comment (jerryshao) [inline catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java:156]: The code here is not robust enough, you need to have a check mechanism to make sure client pool is really closed beforehand.

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java [hunk 2]
```diff
@@ -65,7 +65,18 @@ public class CachedClientPool implements ClientPool<IMetaStoreClient, TException
     this.clientPoolCache =
         Caffeine.newBuilder()
             .expireAfterAccess(evictionInterval, TimeUnit.MILLISECONDS)
-            .removalListener((ignored, value, cause) -> ((HiveClientPool) value).close())
+            .removalListener(
+                (ignored, value, cause) -> {
+                  // Set the class loader to the isolated class loader.
+                  ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+                  Thread.currentThread()
+                      .setContextClassLoader(CachedClientPool.class.getClassLoader());
```
Comment (FANNG1) [inline catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java:73]: why this works? what's CachedClientPool classloader?

Comment (yuqi1129) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java:73]: ``Java
ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
```

The classloader is `AppClassloader`,  Close Hive-related resource needs the Isolated classloader.

Comment (yuqi1129) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java]: It appears that AppClassloader is working fine here, the root cause is not at this point.

Comment (FANNG1) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java]: I prefer to dig out the root cause to check whether there are  some problems with current Isolated classloader.

Comment (jerryshao) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java]: Yep, we should have a reason to explain why.

Comment (yuqi1129) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java]: I have confirmed the root cause: We need to ensure that the removal listener is called before the isolated class loader is closed, as the code logic in the removal listener is asynchronous.  Please refer to the most recent changes.

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java [hunk 3]
```diff
@@ -65,7 +65,11 @@ public class CachedClientPool implements ClientPool<IMetaStoreClient, TException
     this.clientPoolCache =
         Caffeine.newBuilder()
             .expireAfterAccess(evictionInterval, TimeUnit.MILLISECONDS)
-            .removalListener((ignored, value, cause) -> ((HiveClientPool) value).close())
+            .removalListener(
+                (ignored, value, cause) -> {
+                  // We have closed the HiveClientPool that is manually removed from the cache.
```
Comment (jerryshao) [inline catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java:70]: The comment is really hard to understand...

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java [hunk 4]
```diff
@@ -135,6 +139,9 @@ private static ThreadFactory newDaemonThreadFactory() {
   }
 
   public void close() {
+    // Close all the HiveClientPool instances in the cache first and then shutdown the scheduler and
+    // the class loader.
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java:146]: It's in the `close` method of `CatalogMananger`, please see:
image

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java (hunk 3 above)
Comment (yuqi1129) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java]: Modified and clarified it.

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java [hunk 5]
```diff
@@ -65,7 +65,14 @@ public class CachedClientPool implements ClientPool<IMetaStoreClient, TException
     this.clientPoolCache =
         Caffeine.newBuilder()
             .expireAfterAccess(evictionInterval, TimeUnit.MILLISECONDS)
-            .removalListener((ignored, value, cause) -> ((HiveClientPool) value).close())
+            .removalListener(
+                (ignored, value, cause) -> {
+                  // In fact, we have closed some HiveClientPools manually in the `close` method.
+                  // to avoid closing it repeatedly, we add a judgment that if it has been closed,
```
Comment (jerryshao) [inline catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java:71]: Where is the "judgment" you mentioned here?

 I'm super confused about the comment here. You write lots of words here, but doesn't reflect the code here. If it is not here, please add the comment to the code where it happens.

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jerryshao) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java:146]: If it doesn't happen here, why do you mention "shutdown the classloader" here?

[the remark this answers is not in the document]
Comment (yuqi1129) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java:146]: I see.

**Review by jerryshao — APPROVED** [review]

File: core/src/main/java/com/datastrato/gravitino/utils/ClientPoolImpl.java [hunk 6]
```diff
@@ -89,6 +89,12 @@ protected boolean isConnectionException(Exception exc) {
 
   @Override
   public void close() {
```
Comment (FANNG1) [inline core/src/main/java/com/datastrato/gravitino/utils/ClientPoolImpl.java:91]: Is it possible close will be invoked twice?

Comment (yuqi1129) [reply core/src/main/java/com/datastrato/gravitino/utils/ClientPoolImpl.java:91]: I close it manually by `clientPoolCache.asMap().forEach((key, value) -> value.close());` and the removal listener will close again if time expires

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuqi1129) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/CachedClientPool.java]: Yeah, HiveClientPool should be closed before the closing of the classloader.  As the removal listener is asynchronous, it cannot guarantee this.

**Pull request merged**